File: src/main/scala/AmbientWeather.scala
package smick

import com.twitter.conversions.DurationOps._
import com.twitter.finagle.Http
import com.twitter.finagle.http.RequestBuilder
import com.twitter.util.{ Duration, Future }
import java.net.URL

trait AmbientWeather { self: SmickHome =>
  val ambientWeatherMAC = flag("ambientWeather.mac", "", "Weather station MAC")
  val ambientWeatherAPI = flag("ambientWeather.api", "", "Ambient Weather API Key")
  val ambientWeatherApp = flag("ambientWeather.app", "", "Ambient Weather App Key")
  val ambientWeatherFreq = flag("ambientWeather.freq", 1.minute, "Ambient Weather polling frequency")

  def ambientWeatherLoop(store: Store): Future[Unit] =
    loopIt("ambientWeather", ambientWeatherFreq(), process(store))

  private[this] val url = Lazy[URL] {
    new URL(s"https://api.ambientweather.net/v1/devices/${ambientWeatherMAC()}?apiKey=${ambientWeatherAPI()}&applicationKey=${ambientWeatherApp()}&limit=1")
  }

  private[this] val client = Lazy[HttpSvc] {
    val name = "ambient-weather"
    val _url = url()
    Http.client
      .withTls(_url.getHost)
      .newService(destName(name, _url), name)
  }

  private def process(store: Store): Future[Unit] =
    client()(RequestBuilder().url(url()).buildGet()) flatMap { res =>
      val recs = json.readValue[List[Map[String, Any]]](res.contentString)
      val entries = recs flatMap { raw =>
        val data = raw.get("lastData").map(_.asInstanceOf[Map[String, Any]]).getOrElse(raw)
        val ts = {
          val raw = data("dateutc").asInstanceOf[Long]
          val d = Duration.fromMilliseconds(raw)
          Some(d.inNanoseconds)
        }

        StoreEntry("outdoor_temp", data("tempf"), time = ts) ::
        StoreEntry("outdoor_hum", data("humidity"), time = ts) ::
        StoreEntry("wind_dir", data("winddir"), time = ts) ::
        StoreEntry("wind_gust", data("windgustmph"), time = ts) ::
        StoreEntry("wind_speed", data("windspeedmph"), time = ts) ::
        StoreEntry("solar_radiation", data("solarradiation"), time = ts) ::
        StoreEntry("uv_index", data("uv"), time = ts) ::
        StoreEntry("rain_rate_hourly", data("hourlyrainin"), time = ts) ::
        StoreEntry("rain_event_accum", data("eventrainin"), time = ts) ::
        Nil
      }

      store.write(entries)
    }
}

File: src/pkg/looper/ambientweather.go
package looper

import (
	"context"
	"fmt"
	"log/slog"
	"net/url"
	"strings"
	"time"

	"github.com/spf13/pflag"
	hm "github.com/sprsquish/housemetrics/pkg"
	"github.com/sprsquish/housemetrics/pkg/store"
)

type AmbientWeather struct {
	client *hm.HttpClient
	logger *slog.Logger

	mac    string
	apiKey string
	appKey string

	url *url.URL
}

type device struct {
	MacAddr  string          `json:"macAddress"`
	LastData *weatherReading `json:"lastData"`
}

type weatherReading struct {
	DateUTC     int64   `json:"dateutc"`
	OutdoorTemp float64 `json:"tempf"`
	OutdoorHum  float64 `json:"humidity"`
	WindDir     float64 `json:"winddir"`
	WindGust    float64 `json:"windgustmph"`
	WindSpeed   float64 `json:"windspeedmph"`
	Radiation   float64 `json:"solarradiation"`
	UV          float64 `json:"uv"`
	RainRate    float64 `json:"hourlyrainin"`
	RainEvent   float64 `json:"eventrainin"`
}

func NewAmbientWeather(name string, flags *pflag.FlagSet, logger *slog.Logger, client *hm.HttpClient) hm.Looper {
	a := AmbientWeather{
		client: client,
		logger: logger,
	}

	flags.StringVar(&a.mac, fmt.Sprintf("%s.mac", name), "", "Weather station MAC")
	flags.StringVar(&a.apiKey, fmt.Sprintf("%s.api", name), "", "Ambient Weather API Key")
	flags.StringVar(&a.appKey, fmt.Sprintf("%s.app", name), "", "Ambient Weather App Key")

	return &a
}

func (a *AmbientWeather) Init() {
	urlStr := fmt.Sprintf("https://rt.ambientweather.net/v1/devices")
	url, err := url.Parse(urlStr)
	if err != nil {
		a.logger.Error("failed to start", "err", err)
		return
	}

	q := url.Query()
	q.Set("apiKey", a.apiKey)
	q.Set("applicationKey", a.appKey)
	q.Set("limit", "1")
	url.RawQuery = q.Encode()

	a.url = url
}

func (a *AmbientWeather) Poll(ctx context.Context, store store.Client) error {
	var devices []*device
	if err := a.client.GetJSON(ctx, a.logger, &devices, hm.URLOpt(a.url)); err != nil {
		return err
	}

	for _, d := range devices {
		if !strings.EqualFold(d.MacAddr, a.mac) {
			continue
		}

		r := d.LastData

		ts := time.UnixMilli(r.DateUTC)
		store.Write(ctx, ts, "outdoor_temp", r.OutdoorTemp, nil)
		store.Write(ctx, ts, "outdoor_hum", r.OutdoorHum, nil)
		store.Write(ctx, ts, "wind_dir", r.WindDir, nil)
		store.Write(ctx, ts, "wind_gust", r.WindGust, nil)
		store.Write(ctx, ts, "wind_speed", r.WindSpeed, nil)
		store.Write(ctx, ts, "solar_radiation", r.Radiation, nil)
		store.Write(ctx, ts, "uv_index", r.UV, nil)
		store.Write(ctx, ts, "rain_rate_hourly", r.RainRate, nil)
		store.Write(ctx, ts, "rain_event_accum", r.RainEvent, nil)
	}

	return nil
}
